### PR TITLE
[W-11132253 / W-11091132] - Cancel recurring donation

### PR DESCRIPTION
**Pull request by maurimart**
[[Technical Design Review] W-11132253 - NPSP](https://salesforce.quip.com/PLdcAM3askNg)
[[Technical Design Review] W-11091132 - NPSP](https://salesforce.quip.com/MhvnAiCjaNS6)

# Critical Changes

Add cancel recurring donation handler for modal and method for RD2_ETableController.
Disable recurring donation actions based on rd status.

# Changes

### CODE REVIEW

**Comment by maurimart** [comment]
[W-11132253](https://gus.lightning.force.com/a07EE00000wrbMXYAY)

**Review by gerardorodriguezsf — CHANGES_REQUESTED** [review]
- Update apex method to use the recurring donation record.
- Update method to just upsert the record.
- Check CRUD permissions.
- Update Cancel LWC to get the Recurring Donation record, update the Status, and send to apex method to be upserted.
- Prevent duplicated actions due to the rendered actions.

**Comment by maurimart** [comment]
> * Prevent

Changes applied!!!

**Review by gerardorodriguezsf — CHANGES_REQUESTED** [review]
Missing exception for missing permissions:
`} else {
                throw new UTIL_Permissions.InsufficientPermissionException(Label.commonInsufficientPermissions);
            }`

**Review by gerardorodriguezsf — APPROVED** [review]

**Comment by maurimart** [comment]
[W-11091132](https://gus.lightning.force.com/a07EE00000wFnf5YAC)

File: force-app/main/default/classes/RD2_ETableController.cls [hunk 1]
```diff
@@ -105,6 +105,23 @@ public with sharing class RD2_ETableController {
             throw new AuraHandledException(ex.getMessage());
         }
     }
+
+    /**
+     * @description returns a list of processed recurring donations
```
Comment (lparrott) [inline force-app/main/default/classes/RD2_ETableController.cls:110]: Description and return are wrong here.

File: force-app/main/default/labels/CustomLabels.labels-meta.xml [hunk 2]
```diff
@@ -1849,6 +1849,30 @@
...
+        <shortDescription>Text indicating the Recurring Donation would be cancelled</shortDescription>
+        <value>Your request to stop your Recurring Donation was submitted.</value>
+    </labels>
+    <labels>
+        <fullName>RD2_ElevateRDCancellingMessage</fullName>
+        <categories>Recurring Donation</categories>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>Text indicating the Recurring Donation would be cancelled</shortDescription>
+        <value>This can take some time. Refresh the page to check the status. Contact us if you need help.</value>
```
Comment (lparrott) [inline force-app/main/default/labels/CustomLabels.labels-meta.xml:1866]: Are these Labels all approved? "Contact us if you need help." doesn't seem like something we'd normally put in a message 🤷

File: force-app/main/default/lwc/rd2RecurringDonation/rd2RecurringDonation.js [hunk 3]
```diff
@@ -299,18 +300,31 @@ export default class RecurringDonationTable extends LightningElement {
...
     getRecurringDonationFields() {
       getData()
         .then((data) => {
-            if (data) {
-                this.data = data;
-                this.data.forEach(element => {
-                  if(element.nextDonation){
-                    element.nexDonationFormatFirstElement = element.nextDonation.split('.')[0] || element.nextDonation;
-                    element.nexDonationFormatSecondElement = element.nextDonation.split('.')[1] || '';
+          console.log(JSON.parse(JSON.stringify(data)));
```
Comment (lparrott) [inline force-app/main/default/lwc/rd2RecurringDonation/rd2RecurringDonation.js:309]: Remove console.log

**Review by lparrott — CHANGES_REQUESTED** [review]

File: force-app/main/default/lwc/rd2RecurringDonation/rd2RecurringDonation.js (hunk 3 above)
Comment (maurimart) [reply force-app/main/default/lwc/rd2RecurringDonation/rd2RecurringDonation.js]: removed, thanks!

File: force-app/main/default/classes/RD2_ETableController.cls (hunk 1 above)
Comment (maurimart) [reply force-app/main/default/classes/RD2_ETableController.cls:110]: Changed!

File: force-app/main/default/labels/CustomLabels.labels-meta.xml (hunk 2 above)
Comment (maurimart) [reply force-app/main/default/labels/CustomLabels.labels-meta.xml:1866]: @lparrott the labels are defined in this document https://salesforce.quip.com/kRYaAjEE8aHr the text is as described.

File: force-app/main/default/lwc/rd2RecurringDonation/rd2RecurringDonation.html [hunk 4]
```diff
@@ -129,36 +129,36 @@
...
-                        <td key={item.nextDonation} data-id={item.id} class={tdClasses}  data-label={labels.nextDonation}>
+                        <td key={item.nextDonation} data-id={item.recurringDonation.Id} class={tdClasses}  data-label={labels.nextDonation}>
                             <div class={rowClasses} style={fixedWidth}> 
                                 {item.nexDonationFormatFirstElement} </br>
                                 {item.nexDonationFormatSecondElement}
                             </div>
                         </td>
-                        <td key={item.lastModified} data-id={item.id} class={tdClasses}  data-label={labels.lastModified}>
+                        <td key={item.lastModified} data-id={item.recurringDonation.Id} class={tdClasses}  data-label={labels.lastModified}>
                             <div class={rowClasses} style={fixedWidth} title="Last Modified">
```
Comment (force2b) [inline force-app/main/default/lwc/rd2RecurringDonation/rd2RecurringDonation.html:148]: hardcoded text (though maybe this was fixed in another PR?)

Comment (maurimart) [reply force-app/main/default/lwc/rd2RecurringDonation/rd2RecurringDonation.html:148]: Yes, this was fixed in another PR, would we still need to fix it here?

Comment (force2b) [reply force-app/main/default/lwc/rd2RecurringDonation/rd2RecurringDonation.html:148]: If it's a small change then I recommend it, especially if this is going to be UI tested before merging. It also ensures that future merges don't accidentally overwrite the fix to the label with hardcoded text.

Comment (maurimart) [reply force-app/main/default/lwc/rd2RecurringDonation/rd2RecurringDonation.html]: Fixed!!

File: force-app/main/default/lwc/stopRecurringDonationModal/stopRecurringDonationModal.html [hunk 5]
```diff
@@ -15,8 +15,8 @@ <h1 id="modal-heading-01" class="slds-modal__title slds-hyphenate">{labels.stopR
             <p class="slds-p-around_xx-large slds-align_absolute-center">The nonprofit won't process any future donations for this Recurring Donation.</p>
           </div>
           <div class="slds-modal__footer">
-            <button class="slds-button slds-button_neutral" aria-label="Cancel and close" onclick={closeModal}>Cancel</button>
-            <button class="slds-button slds-button_brand" onclick={closeModal}>{labels.stopRecurringDonation}</button>
+            <button class="slds-button slds-button_neutral" aria-label="Cancel and close" onclick={closeModal}>{labels.commonClose}</button>
```
Comment (force2b) [inline force-app/main/default/lwc/stopRecurringDonationModal/stopRecurringDonationModal.html:18]: hardcoded text here

**Review by force2b — CHANGES_REQUESTED** [review]

File: force-app/main/default/lwc/stopRecurringDonationModal/stopRecurringDonationModal.html (hunk 5 above)
Comment (maurimart) [reply force-app/main/default/lwc/stopRecurringDonationModal/stopRecurringDonationModal.html:18]: Changed!

**Review by lparrott — APPROVED** [review]

File: force-app/main/default/labels/CustomLabels.labels-meta.xml [hunk 6]
```diff
@@ -6152,6 +6152,14 @@ To check your GAU Allocation settings, go to Donations | GAU Allocations.</value
         <shortDescription>Close button label</shortDescription>
         <value>Close</value>
     </labels>
+    <labels>
+        <fullName>commonCancelAndClose</fullName>
+        <categories>Button</categories>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>Cancel and close label</shortDescription>
+        <value>Cancel and close</value>
```
Comment (force2b) [inline force-app/main/default/labels/CustomLabels.labels-meta.xml:6161]: Should this be "Cancel and Close"?
Has this label been reviewed by docs?

Comment (maurimart) [reply force-app/main/default/labels/CustomLabels.labels-meta.xml:6161]: yes, this corresponds to the modal's close icon, as defined in the slds documentation

**Pull request merged**